pysrc/receive_message.py

```diff
@@ -51,20 +52,23 @@ def decode_message(self, conn):
                     message = conn.recv(1024).decode(errors="ignore")
             print("message received")
 
+            data_map = json.loads(message_buf[:length_same])
 
-            data_map = ast.literal_eval(message_buf[:length_same])
             message_type = data_map["type"]
-            if message_type == "environment_frame":
-                height = data_map["height"]
-                width = data_map["width"]
+            if message_type == "frame":
+                environment = data_map["environment"]
+
+                height = environment["height"]
+                width = environment["width"]
                 # Global_access.define_grid(width, height)   Uncomment when making grid dynamically sized
-                data_map["grid"] = base64.b64decode(data_map["grid"])
-                data_map["grid"] = [data_map["grid"][i: i + 11] for i in range(0, len(data_map["grid"]), 11)]
-                decode_grid(self, data_map["grid"], width, height)
-                decode_organisms(self, data_map["organisms"])
+                environment["grid"] = base64.b64decode(environment["grid"])
+                environment["grid"] = [environment["grid"][i: i + 11] for i in range(0, len(environment["grid"]), 11)]
+                decode_grid(self, environment["grid"], width, height)
+                decode_organisms(self, environment["organisms"])
                 Global_access.new_frame = True
             elif message_type == "organism_data":
                 pass
+            # TODO handle the response message_type == "control" if necessary
```

Comment:
when you send a controls message, I send back my own controls message that confirms what I received and changed my values to. Use this to make sure that your controls are kept in sync